File: 371_Sum_of_Two_Integers/getSum.py
class Solution(object):
    def getSum(self, a, b):
        bits_a = list(bin(a)[2:][::-1])
        bits_b = list(bin(b)[2:][::-1])
        length = max(len(bits_a), len(bits_b)) + 1
        while len(bits_a) < length:
            bits_a.append('0')
        while len(bits_b) < length:
            bits_b.append('0')
        print bits_a
        print bits_b
        
        res = []
        carry = 0
        for i in xrange(length):
            bita = int(bits_a[i])
            bitb = int(bits_b[i])
            bit = bita ^ bitb ^ carry
            carry = 1 if bita & bitb or bita & carry or bitb & carry else 0
            res.insert(0, str(bit))
        return int(''.join(res), 2)
        
sol = Solution()
a = -1
b = 2
print bin(a)
print bin(b)
print sol.getSum(a, b)

File: 371_Sum_of_Two_Integers/getSum.cpp
#include <iostream>

using namespace std;

class Solution {
    public:
        int getSum(int a, int b) {
            int sum, carry;
            if(b == 0) return a;
            sum = a ^ b;
            carry = (a & b) << 1;
            return getSum(sum, carry);
        }
};

int main() {
    int a = 3;
    int b = 5;
    Solution sol;
    cout << sol.getSum(a, b) << endl;
}
